boot/bootstate20.go

```diff
@@ -46,10 +48,37 @@ func newBootState20(typ snap.Type, dev snap.Device) bootState {
 	}
 }
 
+// modeenvMu is used to protect sections doing:
+//  * read moddeenv/modify it(/reseal from it)
+//  * write modeenv/seal from it
+// while we might want to release the global state lock as seal/reseal are slow
+// (see Unlocker for that)
+var (
+	modeenvMu     sync.Mutex
+	modeenvLocked int32
+)
+
+func modeenvLock() {
+	modeenvMu.Lock()
+	atomic.AddInt32(&modeenvLocked, 1)
+}
+
+func modeenvUnlock() {
+	atomic.AddInt32(&modeenvLocked, -1)
+	modeenvMu.Unlock()
+}
+
+func isModeeenvLocked() bool {
+	return atomic.LoadInt32(&modeenvLocked) == 1
+}
+
 func loadModeenv() (*Modeenv, error) {
+	if !isModeeenvLocked() {
```

Comment:
Shouldn't bootStateUpdate20.commit() do also this check? In the end it is an atomic write file, but if the pattern of usage is going to be read current, find out something, write new values, it would make sense to check that during this full process modeenv is set.

Reply:
I didn't put it there because that calls releaseKeyToModeenv which also does the check. But it doesn't hurt to check twice also and before the write